File: macOS/BotSpeaker/SpeechTextChunker.swift
import Foundation

struct SpeechChunkPlan: Sendable {
    let text: String
    let sourceRange: NSRange
}

enum SpeechTextChunker {
    private static let targetCharacterCount = 420
    private static let minimumCharacterCount = 240
    private static let maximumCharacterCount = 650

    static func chunks(for source: String) -> [SpeechChunkPlan] {
        guard let contentRange = source.rangeOfCharacter(from: .whitespacesAndNewlines.inverted) else {
            return []
        }

        let upperBound = source.rangeOfCharacter(
            from: .whitespacesAndNewlines.inverted,
            options: .backwards
        )!.upperBound
        let fullContentRange = contentRange.lowerBound..<upperBound
        var rawRanges: [Range<String.Index>] = []
        var cursor = fullContentRange.lowerBound

        while cursor < fullContentRange.upperBound {
            let remaining = source.distance(from: cursor, to: fullContentRange.upperBound)
            if remaining <= maximumCharacterCount {
                rawRanges.append(cursor..<fullContentRange.upperBound)
                break
            }

            let maximumEnd = source.index(
                cursor,
                offsetBy: maximumCharacterCount,
                limitedBy: fullContentRange.upperBound
            ) ?? fullContentRange.upperBound
            let minimumEnd = source.index(
                cursor,
                offsetBy: minimumCharacterCount,
                limitedBy: maximumEnd
            ) ?? maximumEnd
            let targetEnd = source.index(
                cursor,
                offsetBy: targetCharacterCount,
                limitedBy: maximumEnd
            ) ?? maximumEnd

            var sentenceBreaks: [String.Index] = []
            var wordBreaks: [String.Index] = []
            var index = cursor

            while index < maximumEnd {
                let character = source[index]
                let next = source.index(after: index)
                if character.isWhitespace {
                    wordBreaks.append(next)
                }
                if character == "." || character == "!" || character == "?" || character == "\n" {
                    sentenceBreaks.append(next)
                }
                index = next
            }

            let sentenceEnd = bestBreak(
                in: sentenceBreaks,
                minimum: minimumEnd,
                target: targetEnd,
                maximum: maximumEnd
            )
            let wordEnd = bestBreak(
                in: wordBreaks,
                minimum: minimumEnd,
                target: targetEnd,
                maximum: maximumEnd
            )
            let end = sentenceEnd ?? wordEnd ?? maximumEnd
            rawRanges.append(cursor..<end)

            cursor = end
            while cursor < fullContentRange.upperBound, source[cursor].isWhitespace {
                cursor = source.index(after: cursor)
            }
        }

        return rawRanges.enumerated().map { index, range in
            let text = String(source[range]).trimmingCharacters(in: .whitespacesAndNewlines)
            let sourceRange = NSRange(range, in: source)
            return SpeechChunkPlan(text: text, sourceRange: sourceRange)
        }
        .filter { !$0.text.isEmpty }
    }

    private static func bestBreak(
        in candidates: [String.Index],
        minimum: String.Index,
        target: String.Index,
        maximum: String.Index
    ) -> String.Index? {
        let valid = candidates.filter { $0 >= minimum && $0 <= maximum }
        return valid.first(where: { $0 >= target }) ?? valid.last
    }
}

File: Windows/BotSpeaker/SpeechTextChunker.cs
namespace BotSpeaker;

/// <summary>A contiguous slice of the source text, in UTF-16 code units.</summary>
public readonly record struct TextSpan(int Location, int Length)
{
    public int End => Location + Length;
}

public sealed record SpeechChunkPlan(string Text, TextSpan SourceRange, string? PreviousText, string? NextText);

/// <summary>
/// Splits a long script into ElevenLabs-sized requests, preferring sentence breaks,
/// then word breaks, between the minimum and maximum chunk sizes. Mirrors the macOS
/// SpeechTextChunker so the two apps produce identical cache keys for identical text.
/// </summary>
public static class SpeechTextChunker
{
    private const int TargetCharacterCount = 420;
    private const int MinimumCharacterCount = 240;
    private const int MaximumCharacterCount = 650;
    private const int ContextCharacterCount = 300;

    public static List<SpeechChunkPlan> Chunks(string source)
    {
        var plans = new List<SpeechChunkPlan>();
        int contentStart = 0;
        while (contentStart < source.Length && char.IsWhiteSpace(source[contentStart])) contentStart++;
        if (contentStart >= source.Length) return plans;

        int contentEnd = source.Length;
        while (contentEnd > contentStart && char.IsWhiteSpace(source[contentEnd - 1])) contentEnd--;

        var ranges = new List<(int Start, int End)>();
        int cursor = contentStart;

        while (cursor < contentEnd)
        {
            int remaining = contentEnd - cursor;
            if (remaining <= MaximumCharacterCount)
            {
                ranges.Add((cursor, contentEnd));
                break;
            }

            int maximumEnd = Math.Min(cursor + MaximumCharacterCount, contentEnd);
            int minimumEnd = Math.Min(cursor + MinimumCharacterCount, maximumEnd);
            int targetEnd = Math.Min(cursor + TargetCharacterCount, maximumEnd);

            var sentenceBreaks = new List<int>();
            var wordBreaks = new List<int>();
            for (int index = cursor; index < maximumEnd; index++)
            {
                char character = source[index];
                if (char.IsWhiteSpace(character)) wordBreaks.Add(index + 1);
                if (character is '.' or '!' or '?' or '\n') sentenceBreaks.Add(index + 1);
            }

            int? end = BestBreak(sentenceBreaks, minimumEnd, targetEnd, maximumEnd)
                ?? BestBreak(wordBreaks, minimumEnd, targetEnd, maximumEnd);
            int chunkEnd = end ?? maximumEnd;
            ranges.Add((cursor, chunkEnd));

            cursor = chunkEnd;
            while (cursor < contentEnd && char.IsWhiteSpace(source[cursor])) cursor++;
        }

        foreach (var (start, end) in ranges)
        {
            var text = source[start..end].Trim();
            if (text.Length == 0) continue;
            plans.Add(new SpeechChunkPlan(
                text,
                new TextSpan(start, end - start),
                ContextBefore(start, source),
                ContextAfter(end, source)));
        }
        return plans;
    }

    private static int? BestBreak(List<int> candidates, int minimum, int target, int maximum)
    {
        var valid = candidates.Where(c => c >= minimum && c <= maximum).ToList();
        if (valid.Count == 0) return null;
        foreach (var candidate in valid)
        {
            if (candidate >= target) return candidate;
        }
        return valid[^1];
    }

    private static string? ContextBefore(int index, string source)
    {
        if (index <= 0) return null;
        int start = Math.Max(index - ContextCharacterCount, 0);
        var value = source[start..index].Trim();
        return value.Length == 0 ? null : value;
    }

    private static string? ContextAfter(int index, string source)
    {
        if (index >= source.Length) return null;
        int end = Math.Min(index + ContextCharacterCount, source.Length);
        var value = source[index..end].Trim();
        return value.Length == 0 ? null : value;
    }
}
